src/lib/auth.ts

```diff
@@ -232,11 +237,36 @@ export const authenticate = async (
   const [rawFirebaseJWT, firebaseJWT] = firebaseJWTResult;
   const { unitCode: newUnitCode, problemOrdinal: newProblemOrdinal } = problemIdResult;
 
-  const authenticatedUser = user_type === "learner"
-                              ? classInfo.students.find(student => student.id === uidAsString)
-                              : classInfo.teachers.find(teacher => teacher.id === uidAsString);
+  let fullName: string;
+  let authenticatedUser: StudentUser | TeacherUser | ResearcherUser | undefined = undefined;
+  switch (user_type) {
+    case "learner":
+      authenticatedUser = classInfo.students.find(student => student.id === uidAsString);
+      break;
+    case "teacher":
+      authenticatedUser = classInfo.teachers.find(teacher => teacher.id === uidAsString);
+      break;
+    case "researcher":
+      fullName = `${portalJWT.first_name} ${portalJWT.last_name}`;
+      authenticatedUser = {
+        type: "researcher",
+        id: uidAsString,
+        portal: portalHost,
+        firstName: portalJWT.first_name,
+        lastName: portalJWT.last_name,
+        fullName,
+        className: classInfo.name,
+        initials: initials(fullName) as string,
+        classHash: classInfo.classHash,
+        offeringId: portalService.offeringId
+      };
```

Comment:
I'd guess this approach is going to cause some issues in the runtime when it assumes the user is in class somehow. However I still think is the right approach and we'll just have to fix the issues as they come up.